README
This is bsd-finger-0.17 for Linux.

This package updates bsd-finger-0.16.

If you're reading this off a CD, go right away and check the net
archives for later versions and security fixes. As of this writing the
home site for NetKit is
	ftp://ftp.uk.linux.org/pub/linux/Networking/netkit

Contents: 
	finger		Program for printing user information
	fingerd		Daemon for remote finger access
	
Requires:
	Working compiler, libc, and kernel.

Security:
	bsd-finger-0.17 contains no new security fixes.

	bsd-finger-0.16 fixes some possible denial of service attacks
	against fingerd.

	bsd-finger-0.10 fixed a denial of service situation where
	users' .plan or .project files are named pipes.

	The NetKit-0.09 and earlier versions of this code fixed a
	number of now well-known security problems. Please don't use
	older versions.

Note:
	If you are using the finger daemon from this package with a 
	custom finger client, rather than the finger client in this
	package, you will need to update your client to send carriage
	returns (CR, or '\r' in C) before line feeds (LF, or '\n' in
	C) if the finger client's standard output is a socket.

	This is because as of bsd-finger-0.15, finger probes this
	condition and sends CRs itself instead of expecting fingerd
	to make an extra copy of all the data through a pipe just to
	add CRs in.

	Ignoring this circumstance and always sending LF instead of
	CR/LF will in most cases work, but is not RFC-compliant.

Installation:
	Do "./configure --help" and decide what options you want. The
	defaults should be suitable for most Linux systems. Then run
	the configure script.

	Do "make" to compile.
	Then (as root) do "make install".

	Save a backup copy of any mission-critical program in case the
	new one doesn't work, and so forth. We warned you.

	If you get gcc warnings from files in /usr/include, they are
	due to problems in your libc, not netkit. (You may only see
	them when compiling netkit because netkit turns on a lot of
	compiler warnings.)

DEC CC:
	The DEC compiler for the Alpha is now freely available. This
	is a much better compiler with gcc, that is, it generates much
	better code. If you have the DEC compiler, you can explicitly
	use the DEC compiler instead of gcc by configuring like this:

		./configure --with-c-compiler=ccc

	It is known to generate spurious warnings on some files. Also,
	some headers from some versions of glibc confuse it; that may
	prevent	netkit from working. Other problems should be reported
	as bugs.

Bugs:
	Please make sure the header files in /usr/include match the
	libc version installed in /lib and /usr/lib. If you have weird
	problems this is the most likely culprit.

	Also, before reporting a bug, be sure you're working with the
	latest version.

	If something doesn't compile for you, fix it and send diffs.
	If you can't, send the compiler's error output.

	If it compiles but doesn't work, send as complete a bug report as 
	you can. Patches and fixes are welcome, as long as you describe 
	adequately what they're supposed to fix. Please, one patch per
	distinct fix. Please do NOT send the whole archive back or
	reindent the source.

	Be sure to send all correspondence in e-mail to the netkit address.
	Postings to netnews or mailing lists will not be seen due to the 
	enormous volume. Also, anything that doesn't get filed in the bug
	database is quite likely to end up forgotten.

	Please don't report known bugs (see the BUGS file(s)) unless you
	are including fixes. :-)

	Mail should be sent to: netbug@ftp.uk.linux.org


Early in April 2000, a hacker broke into the machine that was hosting
the netkit bug database for me and trashed it. Unfortunately, it seems
backups hadn't gotten done for a while, so three months of mail (since
mid-January) was lost. So, if you sent something and didn't hear back,
or you sent something, heard back, but the changes failed to appear in
this release (unlikely but possible) - please resend.

Please see http://www.hcs.harvard.edu/~dholland/computers/netkit.html
if you are curious why it was so long between the 0.10 and 0.16 releases.

Future plans for netkit maintenance are still up in the air, but in the
meantime new releases will still appear from time to time. I don't have
a whole lot of cycles to spare to work on netkit, so things are likely
to continue to be fairly slow.

David A. Holland
23 July 2000


ChangeLog
22-Jul-2000:
	Document ~/.pgpkey. (Herbert Xu, herbert@gondor.apana.org.au)

18-Dec-1999:
	Add support for ~/.nofinger files.

14-Dec-1999:
	bsd-finger-0.16 is released.

12-Dec-1999:
	Withdrew -R option; finger can figure out on its own if it needs
	  to emit CRs.
	Also, make printing of pts/* ttys work right.

14-Sep-1999:
	Rewrote output handling of finger to deal with 8-bit characters 
	  correctly. Also fixed what I think was some uninitialized data
	  displaying ":0" utmp entries.
	Backed out patch 1.15 of finger/util.c because it was completely
	  wrong. Fixed gecos processing _again_.
	Added -R option to finger that causes it to emit CR before every
	  LF. Changed fingerd to use it. This means fingerd no longer 
	  needs to fork.

1-Aug-1999:
	Complete y2k and y2038 audit.

31-Jul-1999:
	Redid makefiles/config stuff for new confgen version.

23-Sep-1997:
	fingerd should now refuse to run as root and bail if it can't find
	"nobody". 

22-Sep-1997:
	Fix finger to not destroy gethostbyname() results with
	getservbyname(). 

02-Aug-1997:
	Fix fingerd to complain if it can't switch from root.

12-Jun-1997:
	bsd-finger-0.10 released.

08-Jun-1997:
	More adjustments for glibc.

13-May-1997:
	Do dynamic column width formatting in short format printout. (HJ Lu)

05-Apr-1997:
	Added configure script to generate MCONFIG.
	glibc fixes from HJ Lu.
	Display .pgpkey file in finger (Edward S. Marshall,
	  emarshal@common.net)
	Better utmp handling.

08-Mar-1997:
	Split from full NetKit package. 
	Generated this change log from NetKit's.

07-Mar-1997:
	Fixed finger to reject .plans that aren't regular files. (Thomas
	  Roessler, roessler@sobolev.rhein.de)

01-Feb-1997:
	Fixed finger's timezone handling. Now gets daylight time right.

29-Dec-1996
	NetKit-0.09 released.
	Fix doc bugs in fingerd.
	Fingerd accepts -p as equivalent to -L for compatibility with
	  "fingerd-extlog.tar.gz".
	Assorted alpha/glibc patches. (Erik Troan, ewt@redhat.com)
	Assorted bug fixes from Debian. (Peter Tobias, 
	  tobias@et-inf.fho-emden.de)
	Hardened programs against DNS h_length spoofing attacks.
	Use inet_aton() everywhere instead of inet_addr().
	Fixed bug in finger's processing of &'s in full names that probably has
	  security implications.
	Finger doesn't follow symlinks now.

22-Aug-1996
	NetKit-B-0.08 released.
	fingerd uses fork instead of system for running uptime (with -w flag).
	fingerd supports -L flag to run alternate finger client.
	(almost) everything now compiles with lots of warnings turned on.
	fingerd now accepts a new -u flag to reject "finger @host" queries.
	fix bug in finger that prevented matching full names
	  (fix from James Jones <james@richland.cc.il.us>.)

25-Jul-1996
	NetKit-B-0.07A released.

23-Jul-1996
	NetKit-B-0.07 released.
	Integrated a collection of patches that had been lurking on the net,
	  including the 256-ptys support for telnetd and passive mode ftp.
	Major security fixes, including to fingerd, lpr, rlogin, rsh, talkd, 
	  and telnetd. Do *not* use the sliplogin from earlier versions of this
	  package, either.
	Several of the daemons support better logging.
	Much of the code builds without libbsd.a or bsd includes.
	Massive code cleanup. Almost everything compiles clean with gcc
	  -Wall now. rusers and rusersd do not; patches to rpcgen to fix
	  this would be appreciated if anyone feels like it.
	New maintainer:  David A. Holland, dholland@hcs.harvard.edu

date not known
	NetKit-B-0.06 released.
	Added a missing argument in a printf to the finger source
	  code. Phil Edge
	"finger -l" prints the standard timezone when it should print the
	  daylight savings timezone. tzname[0] is standard timezone and
	  tzname[1] is daylight savings timezone.  This patch is from Phil
	  Edge. He wasn't too sure about this patch, and I haven't looked
	  into it. Anyone who really knows about this stuff? Look into
	  lprint.c in the finger source. I changed all "1 - daylight" into
	  "daylight"... You could also verify all Linux-specific changes
	  to timezone things. Other parts look also like duplicate work.
	  (Maybe even wrong...)

date not known
	NetKit-B-0.05 released.
	Fixed finger/util.c to cope with XDM login entry in utmp (Leonard
	  Zubkoff)

date not known
	NetKit-B-0.04 released.

date not known
	NetKit-B-0.03 released.
	Fixed fingerd to not use 'getdomainname' to get the FQDN.


fingerd/fingerd.c
/*
 * Copyright (c) 1983 The Regents of the University of California.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

char copyright[] =
  "@(#) Copyright (c) 1983 The Regents of the University of California.\n"
  "All rights reserved.\n";

/* 
 * from: @(#)fingerd.c	5.6 (Berkeley) 6/1/90"
 */
char rcsid[] = 
  "$Id: fingerd.c,v 1.23 1999/12/12 18:46:28 dholland Exp $";

#include <pwd.h>
#include <grp.h>
#include <netdb.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <signal.h>
#include <syslog.h>
#include <getopt.h>
#include <netinet/in.h>
#include <sys/utsname.h>
#include <sys/wait.h>

#include "pathnames.h"
#include "../version.h"

#define	ENTRIES	50
#define WS " \t\r\n"

/* These are used in this order if the finger path compiled in doesn't work. */
#define _ALT_PATH_FINGER_1 "/usr/local/bin/finger"
#define _ALT_PATH_FINGER_2 "/usr/ucb/finger"
#define _ALT_PATH_FINGER_3 "/usr/bin/finger"

static
void
fatal(const char *msg, int use_errno, int tolog, int toclient)
{
	const char *err = "";
	const char *sep = "";
	if (use_errno) {
		err = strerror(errno);
		sep = ": ";
	}
	if (tolog) syslog(LOG_ERR, "%s%s%s\n", msg, sep, err);
	if (toclient) fprintf(stderr, "fingerd: %s%s%s\r\n", msg, sep, err);
	else fprintf(stderr, "fingerd: Internal error\r\n");
	exit(1);
}

static
void
timeout(int sig)
{
	(void)sig;
	errno = ETIMEDOUT;
	fatal("Input timeout", 0, 1, 1);
}


int
main(int argc, char *argv[])
{
#if 0
	FILE *fp;
	int p[2], ch;
	pid_t pid;
#endif
	int ca;
	const char *av[ENTRIES + 1];
	const char **avy;
	char *const *avx;
	char line[1024];
	int welcome = 0, heavylogging = 0, nouserlist = 0;
	int patience = 60, forwarding = 0;
	int k, nusers;
	char *s, *t;
	const char *fingerpath = NULL;
	struct sockaddr_in sn;
	socklen_t sval = sizeof(sn);


	if (getpeername(0, (struct sockaddr *) &sn, &sval) < 0) {
		fatal("getpeername", 1, 0, 1);
	}

	openlog("fingerd", LOG_PID, LOG_DAEMON);

	if (!getuid() || !geteuid()) {
		struct passwd *pwd = getpwnam("nobody");
		if (pwd) {
			initgroups(pwd->pw_name, pwd->pw_gid);
			setgid(pwd->pw_gid);
			setuid(pwd->pw_uid);
		}
		seteuid(0);   /* this should fail */
		if (!getuid() || !geteuid()) {
			fatal("setuid: couldn't drop root", 0, 1, 0);
		}
	}
	/*endpwent();  -- is it safe to put this here? */

	opterr = 0;
	while ((ca = getopt(argc, argv, "wlL:p:uft:h?")) != EOF) {
		switch(ca) {
		  case 'w':
			welcome = 1;
			break;
		  case 'l':
		        heavylogging = 1;
			break;
		  case 'L': 
		  case 'p':
		        fingerpath = optarg;
			break;
		  case 'u':
		        nouserlist = 1;
			break;
		  case 'f':
		        forwarding = 1;
			break;
		  case 't':
		        patience = atoi(optarg);
			break;
		  case '?':
		  case 'h':
		  default:
			syslog(LOG_ERR, "usage: fingerd [-wulf]"
					"[-pL /path/finger] [-t timeout]");
			exit(1);
		}
	}
	argc -= optind;
	argv += optind;

	/*
	 * Hang up after a while so people can't DoS by leaving lots of
	 * open sockets about.
	 */
	if (patience != 0) {
		signal(SIGALRM, timeout);
		alarm(patience);
	}
	if (!fgets(line, sizeof(line), stdin)) {
		fatal("Client hung up - probable port-scan", 0, 1, 0);
	}

	if (welcome) {
		char buf[256];
		struct hostent *hp;
		struct utsname utsname;

		uname(&utsname);
		gethostname(buf, sizeof(buf));
		if ((hp = gethostbyname(buf))) {
			/* paranoia: dns spoofing? */
			strncpy(buf, hp->h_name, sizeof(buf));
			buf[sizeof(buf)-1] = 0;
		}
		printf("\r\nWelcome to %s version %s at %s !\r\n\n",
				utsname.sysname, utsname.release, buf);
		fflush(stdout);
		switch (fork()) {
		 case -1: /* fork failed, oh well */
		     break;
		 case 0: /* child */
		     execl(_PATH_UPTIME, _PATH_UPTIME, NULL);
		     _exit(1);
		 default: /* parent */
		     wait(NULL);
		     break;
		}
		fflush(stdout);
		printf("\r\n");
		fflush(stdout);
	}

	k = nusers = 0;
	av[k++] = "finger";
	for (s = strtok(line, WS); s && k<ENTRIES; s = strtok(NULL, WS)) {
		/* RFC742: "/[Ww]" == "-l" */
		if (!strncasecmp(s, "/w", 2)) memcpy(s, "-l", 2);
		if (!forwarding) {
		    t = strchr(s, '@');
		    if (t) {
			    fprintf(stderr,
				    "fingerd: forwarding not allowed\r\n");
			    syslog(LOG_WARNING, "rejected %s\n", s);
			    exit(1);
		    }
		}
		if (heavylogging) {
		    if (*s=='-') syslog(LOG_INFO, "option %s\n", s);
		    else syslog(LOG_INFO, "fingered %s\n", s);
		}
		av[k++] = s;
		if (*s!='-') nusers++;
	}
	av[k] = NULL;
	if (nusers==0) {
		/* finger @host */
		if (nouserlist) {
			syslog(LOG_WARNING, "rejected finger @host\n");
			printf("Please supply a username\r\n");
			return 0;
		}
		if (heavylogging) syslog(LOG_INFO, "fingered @host\n");
	}

/* Yay! we don't need to do this any more - finger does it for us */
#if 0
	if (pipe(p) < 0) {
		fatal("pipe", 1, 1, 0);
	}

	pid = fork();
	if (pid<0) {
		fatal("fork", 1, 1, 0);
	}
	if (pid==0) {
		/* child */
		close(p[0]);
		dup2(p[1], 1);
		if (p[1]!=1) close(p[1]);
#endif
		/*
		 * execv() takes (char *const *), because (char const *const *)
		 * doesn't work right in C (only in C++). C9x might fix this
		 * if we're lucky. In the meantime we need to defeat the type
		 * system to avoid warnings.
		 */
		avy = av;
		/*avx = avy;*/
		memcpy(&avx, &avy, sizeof(avx));

		if (fingerpath) execv(fingerpath, avx);
		execv(_PATH_FINGER, avx);
		execv(_ALT_PATH_FINGER_1, avx);
		execv(_ALT_PATH_FINGER_2, avx);
		execv(_ALT_PATH_FINGER_3, avx);
		syslog(LOG_ERR, "Finger program not found\n");
		exit(1);
#if 0
	}
	/* parent */
	close(p[1]);

	/* convert \n to \r\n. This should be an option to finger... */
	fp = fdopen(p[0], "r");
	if (!fp) {
		fatal("fdopen", 1, 1, 0);
	}

	while ((ch = getc(fp)) != EOF) {
		if (ch == '\n')	putchar('\r');
		putchar(ch);
	}
	return 0;
#endif
}
